### ISSUE TITLE
Doors get automatically bypassed during arming

### ISSUE THREAD

**Issue by timaseth** [nielsfaber/alarmo#786]
### Alarmo Version

v1.9.10

### HA Version

2023.8

### Bug description

Setup is a simple ARM_NIGHT mode with a few door and window sensors.
Now, irrespective of the settings I try on any door sensor, all open doors get automatically bypassed while arming. These can be 1, 2 or even 3 doors. 
Tried applying window settings to the doors but that did not stop them getting bypassed.

### Steps to reproduce

Ensured ALL sensor settings for every door were Off. Even setup the doors as Windows.
At some random times, any particular door may give an expected arming error but I am unable to reliably reproduce it.

### Relevant log output

```shell
HA State:
code_format: null
changed_by: amit
code_arm_required: true
arm_mode: armed_night
open_sensors: null
bypassed_sensors: binary_sensor.sidegatestate, binary_sensor.basementdoorstate, binary_sensor.bluedoorstate
delay: null
icon: mdi:shield-home
friendly_name: bd1
supported_features: 12
```

**Comment by nielsfaber** [comment]
How do you arm the alarm?
Are you using the alarmo-card for this, or otherwise?

**Comment by timaseth** [comment]
Thanks for such a quick reply!
Yes, the special Alarmo Lovelace card.

**Comment by nielsfaber** [comment]
Alarmo will only bypass sensors in case:
1. this requested in the arming command; OR
2. you have given permission in the configuration of the sensor

AND (in both cases) the sensor is active/on when the alarm is armed (=after exit delay finishes).

I understand you use the alarmo-card and you leave the 'bypass open sensors' unchecked.
Still, to exclude some potential bug in the card (accidentally setting the property), I propose that you try to arm the alarm via the Developer Tools -> Services panel. 
You can use the `alarmo.arm` service and only provide the entity id, mode and pin code.

The second scenario would be that the 'bypass automatically' setting is still enabled for the affected sensors (even though the GUI tells otherwise).
By default this setting is not set. So I propose to remove the sensor(s) from alarmo (click the 'remove' button on the sensor configuration page) and add them again (without touching any settings).

Another thing to consider is checking your MQTT traffic. If you have MQTT enabled for Alarmo, the alarm can be armed via this method as well. Possibly some command is being sent on the background with the bypass_open_sensors setting applied. 

I cannot think of anything else to check. You seem to be the only user who suffers this issue, so at this point it is more likely to be related to your usage than Alarmo itself.

I have also not been able to reproduce the issue following the steps you provided.
Some final remark regarding the steps you provided:
> Ensured ALL sensor settings for every door were Off. Even setup the doors as Windows.

If a sensor remains off during arming (until the exit delay has finished), it would not be bypassed by alarmo _by design_.
If you are still convinced this is the case, I would like to see some evidence. This could be for example a screen video in which you demonstrate the arming process as you perform it, while keeping the sensor states + alarm state (and bypassed_sensors property) visible.

**Comment by timaseth** [comment]
Thanks for the detailed reply.
1. I tried using the service call methodology
```
service: alarmo.arm
data:
  mode: night
  code: "XXXX"
  entity_id: alarm_control_panel.bd1
  ```
  But was able to arm with two doors being open
  
<img width="818" alt="image" src="https://github.com/nielsfaber/alarmo/assets/60132339/f184757e-6b26-4eec-8976-51d89626407f">
HA State

`code_format: null
changed_by: amit
code_arm_required: true
arm_mode: armed_night
open_sensors: null
bypassed_sensors: binary_sensor.sidegatestate, binary_sensor.basementdoorstate
delay: null
icon: mdi:shield-home
friendly_name: bd1
supported_features: 12`

Basement Door
<img width="465" alt="image" src="https://github.com/nielsfaber/alarmo/assets/60132339/138e0734-5aca-40d0-9894-1d0131a1b800">
BlueDoor
<img width="466" alt="image" src="https://github.com/nielsfaber/alarmo/assets/60132339/876a08d2-d9e2-4d34-82f1-b5f5b944fa2c">

I will soon revert with feed-back on your other suggestions.

PS: Tried on both Safari and Chrome browsers with the same results.

**Comment by timaseth** [comment]
Step 2.
Removed all sensors.
Just added the 3 doors in question. Tweaked or checked no settings.
<img width="583" alt="image" src="https://github.com/nielsfaber/alarmo/assets/60132339/7ed4f58f-1835-455e-9738-c321ff8d566a">

Here is a small video

https://github.com/nielsfaber/alarmo/assets/60132339/93b7ec70-e3a3-4dd4-a9a6-3cdcfd89d811

Still am where I started.

**Comment by nielsfaber** [comment]
as a last resort, could you copy the contents of the `alarmo.storage` file which is located in your HA configuration directory in subfolder `.storage`? This folder may be hidden depending on your OS, but it does exist.
I am only interested in the section regarding these affected sensors.

**Comment by timaseth** [comment]
Hi Neils,

Here we go!
https://pastebin.com/vfJr5LNV

**Comment by henriklund** [comment]
I've observed a similar behaviour for the past few days. In my case I use 'armed home' on all door / window to trigger a similar behavior (trigger an update of a input_text element used in Lovelace). For all sensors _Trigger when unavailable_ and _Allow open initially_ are enabled. Trigger time is set to one minute.
Previously a door / window could be opened (and alarm triggered) and then left opened past the trigger time (and state returned to armed_home) without any issues. Now I note that when this is occurs and the alarm returns to _armed_home_, the door  / window sensor listed in open_sensors (state _triggered_) will now be included in the _bypassed_sensors_. Closing the door / window does not change this. However triggering alarm again (on a not bypassed door / window sensor) will 'clean up' the _bypassed_sensors_ on any sensors now closed.
I am not sure if this has any relevans, but I do not seem to recall a similar behavior while running Core 2023.8.1 (I upgraded to Core 2023.8.2 Friday afternoon).

**Comment by timaseth** [comment]
Thanks henrik for your inputs.
I upgraded to the latest 2023.8.3 with no improvement.

**Comment by nielsfaber** [comment]
@timaseth 
I copied your configuration to my own HA installation.
Only changes I made:
1. `binary_sensor.basementdoorstate` was replaced by `binary_sensor.backdoor` (existing sensor entity in my own setup)
2. `binary_sensor.bluedoorstate` was replaced by `binary_sensor.fontdoor` (existing sensor entity in my own setup)
3. `binary_sensor.sidegatestate` was replaced by `binary_sensor.studyroom_window` (existing sensor entity in my own setup)
4. Replaced your PIN with my own
5. Replaced notify service with my own

Following your video, I armed the alarm (alarm_control_panel.bd1) with sensors 1 and 3 open, sensor 2 is closed.
The arming succeeds without issues and the open sensors are bypassed:
![image](https://github.com/nielsfaber/alarmo/assets/16861605/b8780869-f8ce-4825-8ad9-9aa5e000b646)

In summary, I finally managed to reproduce the issue you are having, I will investigate what's going on and keep you posted when I have an update.

**Comment by nielsfaber** [comment]
@henriklund 
> Previously a door / window could be opened (and alarm triggered) and then left opened past the trigger time (and state returned to armed_home) without any issues. Now I note that when this is occurs and the alarm returns to armed_home, the door / window sensor listed in open_sensors (state triggered) will now be included in the bypassed_sensors. 

The behaviour you are experiencing is as it is designed (so intended).
The following fragment of the source code describes the behaviour after the trigger time has expired:
https://github.com/nielsfaber/alarmo/blob/0b933d65c44cc630adb46bf8fd1e5f67ef6eed66/custom_components/alarmo/alarm_control_panel.py#L733-L743

The last line shows that open sensors will be bypassed (and exit delay will be skipped), so the alarm is forced back in the armed state. Any open sensors (also sensors which did not cause the previous triggering of the alarm) will be bypassed.
Without this behaviour the arming would fail and the alarm would become disarmed.
This code was last modified in november 2021, so I wouldn't expect any of the behaviour to be recently changed. 

Is this behaviour different to what you would expect?

> Closing the door / window does not change this. However triggering alarm again (on a not bypassed door / window sensor) will 'clean up' the bypassed_sensors on any sensors now closed.

Closing the door before the alarm is re-armed (so during the trigger time) would indeed cause the sensor not to be bypassed later.
In summary, sensors will only be bypassed when needed to make the arming successful. So this may change between arm events.

Is there anything you would like to see changed regarding this?

**Comment by henriklund** [comment]
@nielsfaber 

Interesting...I had the impression that although arming with open sensors (like doors / windows), these would be bypassed if 'Allow open initially' was selected and then subsequently included. I somewhat had a recollection that this was how it worked in the past (although your code shows that it cannot be so).
In my case the problem is that sensors may remain open for longer than the trigger time (for instance keeping windows open during hot summer days). If I understand it correct, closing the sensor (after the automatic re-arming when trigger time expires) leaves the sensor bypassed until next change of state (triggering of another sensor or changing the alarm mode). I would prefer the ability to mark as sensor to be automatically included when closed (and removed from bypassed) or alternatively to be able to manually trigger Alarmo to reevaluate bypassed sensors.

**Comment by timaseth** [comment]
> In summary, I finally managed to reproduce the issue you are having, I will investigate what's going on and keep you posted when I have an update.

Thanks Niels; for allowing me to retain my sanity ;-)
Happy to be of use as and when you deem necessary.
All the best!

**Comment by nielsfaber** [comment]
@timaseth I found out that the behaviour is due to your user setting:
![image](https://github.com/nielsfaber/alarmo/assets/16861605/52e552de-07a4-43a4-9605-47495729a292)

Enabling the 'is override code' for a certain user/code combination, has the effect that the alarm is forcibly armed.
Which means that any sensor which prevents the arming, would have to be bypassed. 

I'm guessing you didn't want this setting applied, so I suggest to uncheck it.
The functionality is intended to be used with a secondary code. If you need to leave the house in a hurry and left a window open, you could enter this code on your keypad to ignore this window.

**Comment by timaseth** [comment]
> @timaseth I found out that the behaviour is due to your user setting

Such a simple solution to what appeared to be a major bug!!
Many thanks @nielsfaber Appreciate the app as well your time!
